Add Podman support and SELinux handling to run_docker_build.sh template

aman-coder03 opened the pull request:
This PR adds Podman support to the `run_docker_build.sh.tmpl` template so feedstocks
can build locally on systems where Docker is not available (e.g., Fedora Silverblue,
RHEL, or environments preferring rootless Podman).

### Key changes
- Added automatic container runtime detection (`docker` → `podman` fallback)
- Replaced hardcoded Docker invocations with `${CONTAINER_RUNTIME_EXE}`
- Added Podman rootless file-permission fix (`podman unshare chown`)
- Added SELinux volume labeling support (`:z`) when enforcing mode is enabled
- Updated volume mounts to correctly append `${VOLUME_SUFFIX}`

### Why this is needed
Many modern Linux distributions ship Podman by default, not Docker. Adding this
logic ensures conda-smithy-generated feedstocks can run local builds seamlessly
across both container runtimes.

This aligns with the improvements made recently in staged-recipes.

isuruf commented on conda_smithy/templates/run_docker_build.sh.tmpl:
```diff
@@ -70,22 +79,38 @@ mkdir -p "$ARTIFACTS"
+
+    # Add SELinux label for proper volume mounting
+    if command -v getenforce &>/dev/null && [ "$(getenforce)" != "Disabled" ]; then
+        VOLUME_SUFFIX=":z"
```
Is this supposed to be
```suggestion
        VOLUME_SUFFIX=",z"
```

isuruf commented on conda_smithy/templates/run_docker_build.sh.tmpl:
```diff
@@ -70,22 +79,38 @@ mkdir -p "$ARTIFACTS"
+# Podman-specific fixes for SELinux and rootless mode
+if [ "${CONTAINER_RUNTIME_EXE}" = "podman" ]; then
+    # Fix file permissions for rootless podman builds
+    podman unshare chown -R 1000:1000 "${ARTIFACTS}"
```
```suggestion
    podman unshare chown -R ${HOST_USER_ID}:${HOST_USER_ID} "${ARTIFACTS}"
```

?

aman-coder03 commented:
@h-vetinari i have added a news entry for this PR under `Added`
please let me know if any changes are needed!

h-vetinari reviewed (commented):
Given that there is already existing templating of `{{ docker.executable }}`, that seems redundant with `${CONTAINER_RUNTIME_EXE}`...? 🤔

h-vetinari commented on conda_smithy/templates/run_docker_build.sh.tmpl:
```diff
@@ -27,14 +29,13 @@ if [[ "${sha:-}" == "" ]]; then
-# Check if docker-machine is being used (normally on OSX) and get the uid from
-# the VM
+
+# Check if docker-machine is being used (normally on OSX)
```
undo these changes.

h-vetinari commented on conda_smithy/templates/run_docker_build.sh.tmpl:
```diff
@@ -70,22 +71,36 @@ mkdir -p "$ARTIFACTS"
 rm -f "$DONE_CANARY"
 
-# Allow people to specify extra default arguments to `docker run` (e.g. `--rm`)
+# Allow people to specify extra default arguments
```
leave this comment as-is

danyeaw commented on conda_smithy/templates/run_docker_build.sh.tmpl:
```diff
@@ -70,22 +71,36 @@ mkdir -p "$ARTIFACTS"
+
+if [ "${CONTAINER_RUNTIME}" = "podman" ]; then
+    # Fix file permissions for rootless podman builds
+    podman unshare chown -R ${HOST_USER_ID}:${HOST_USER_ID} "${ARTIFACTS}"
```
```suggestion
    podman unshare chown -R ${HOST_USER_ID}:${HOST_USER_ID} "${ARTIFACTS}"
    podman unshare chown -R ${HOST_USER_ID}:${HOST_USER_ID} "${RECIPE_ROOT}"
```

aman-coder03 commented:
thanks for the helpful reviews!
i have rebased the branch and applied the requested fixes:
restored the original Docker volume suffix behavior to preserve SELinux compatibility
added the RECIPE_ROOT permission fix for rootless Podman builds
Please let me know if anything else needs adjustment.

h-vetinari commented on conda_smithy/templates/run_docker_build.sh.tmpl:
```diff
@@ -11,6 +11,8 @@ source .scripts/logging_utils.sh
 
 set -xeo pipefail
 
+CONTAINER_RUNTIME="{{ docker.executable }}"
```
FWIW, this is how the (sample) configuration looks that fills in the templating here
https://github.com/conda-forge/conda-smithy/blob/e8e20342ddfb6773103270de5f23b798df55b17c/conda_smithy/data/conda-forge.yml#L71-L75

I'm indifferent about introducing a `CONTAINER_RUNTIME` variable for that (as you do here), or to just keep using `{{ docker.executable }}` throughout the template here.

isuruf replied:
We need a way for a user to use podman for a one-off build.
```
CONTAINER_RUNTIME="${CONTAINER_RUNTIME:-{{ docker.executable }}}"
```
The configuring of `docker.executable` is for a permanent change, but if the user wants to run using podman for a one-off build, then configuring `docker.executable` and rerendering is a hassle. We also do the same logic for `DOCKER_IMAGE`.

h-vetinari reviewed (commented):
basically LGTM, thanks. @isuruf may have opinions about `{{ docker.executable }}` vs `CONTAINER_RUNTIME`

isuruf commented on conda_smithy/templates/run_docker_build.sh.tmpl:
```diff
@@ -11,6 +11,8 @@ source .scripts/logging_utils.sh
 
 set -xeo pipefail
 
+CONTAINER_RUNTIME="${CONTAINER_RUNTIME:-{{ docker.executable }}}"
```
I wonder if we should rename `CONTAINER_RUNTIME` to `DOCKER_EXECUTABLE` since the corresponding options `DOCKER_IMAGE`, `DOCKER_RUN_ARGS` all start with `DOCKER_`

aman-coder03 replied:
Sounds good, I’ll switch to DOCKER_EXECUTABLE everywhere and open a follow-up issue for the broader naming discussion.

aman-coder03 commented:
@isuruf the PR has @h-vetinari 's approval and all checks pass... could you please merge when you get a chance?
